File: cpp/src/camera/webcam_controls.cpp
#include "webcam_controls.hpp"

#include <cstdlib>
#include <iostream>

string WebcamControls::build_command(string attribute, string value) {
  return "v4l2-ctl -d " + device + " --set-ctrl=" + attribute + "=" + value;
}

void run(string command) {
  system(command.c_str());
}

void WebcamControls::enable_automatic_controls() {
  cout << "Enabling automatic webcam controls\n";
  run(build_command("exposure_auto", "3"));
  run(build_command("white_balance_temperature_auto", "1"));
}

void WebcamControls::disable_automatic_controls() {
  cout << "Disabling automatic webcam controls\n";
  run(build_command("exposure_auto", "1"));
  run(build_command("white_balance_temperature_auto", "0"));
}

File: kotlin_old/src/main/kotlin/uk/ac/ic/matterialize/camera/WebcamControls.kt
package uk.ac.ic.matterialize.camera

import uk.ac.ic.matterialize.util.runCommand

class WebcamControls(deviceNumber: Int = 0) {
    private val device = "/dev/video$deviceNumber"
    private val timeout = 5

    init {
        enableAutomaticControls()
    }

    fun disableAutomaticControls() {
        println("Disabling automatic webcam controls")
        "v4l2-ctl -d $device --set-ctrl=exposure_auto=1".runCommand()
        "v4l2-ctl -d $device --set-ctrl=white_balance_temperature_auto=0".runCommand()
    }

    fun enableAutomaticControls() {
        println("Enabling automatic webcam controls")
        "v4l2-ctl -d $device --set-ctrl=exposure_auto=3".runCommand()
        "v4l2-ctl -d $device --set-ctrl=white_balance_temperature_auto=1".runCommand()
    }
}
